lib/web/join_tokens.go

```diff
@@ -55,28 +59,66 @@ type scriptSettings struct {
 	appInstallMode bool
 	appName        string
 	appURI         string
-}
-
-// createTokenRequest is the expected request body of
-// the endpoint to create token
-type createTokenRequest struct {
-	Roles types.SystemRoles `json:"roles"`
+	joinMethod     string
 }
 
 func (h *Handler) createTokenHandle(w http.ResponseWriter, r *http.Request, params httprouter.Params, ctx *SessionContext) (interface{}, error) {
-	var req createTokenRequest
+	var req types.ProvisionTokenSpecV2
 	if err := httplib.ReadJSON(r, &req); err != nil {
-		log.WithError(err).Error("error reading body")
 		return nil, trace.Wrap(err)
 	}
 
 	clt, err := ctx.GetClient()
 	if err != nil {
-		log.WithError(err).Error("error getting client")
 		return nil, trace.Wrap(err)
 	}
 
-	return createJoinToken(r.Context(), clt, req.Roles)
+	// create a new random dynamic token
+	tokenName, err := utils.CryptoRandomHex(auth.TokenLenBytes)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	expires := time.Now().UTC().Add(defaults.NodeJoinTokenTTL)
+
+	// to prevent generation of redundant IAM tokens
+	// we generate a deterministic name for them
+	if req.JoinMethod == types.JoinMethodIAM {
+		tokenName, err = generateIamTokenName(req.Allow)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		// if a token with this name is found, return it
+		// otherwise, go ahead and create it
+		t, err := clt.GetToken(r.Context(), tokenName)
+		if err == nil {
+			return &nodeJoinToken{
```

Comment:
in case there's a hash collision (unlikely) or someone alters the token (more likely), can you sanity check that it includes the expected rule and has the right join method, and just return an error if it doesn't?